target-gen/src/commands/test.rs

```diff
@@ -0,0 +1,218 @@
+
+    let mut loader = session.target().flash_loader();
+    let data = (0..data_size)
+        .into_iter()
+        .map(|n| (n % 256) as u8)
+        .collect::<Vec<_>>();
+    loader.add_data(0x1, &data)?;
+    run_flash_download(&mut session, loader, &progress, true)?;
+    let mut readback = vec![0; data_size as usize];
+    session.core(0)?.read_8(0x1, &mut readback)?;
+    assert_eq!(readback, data);
+
+    println!("{test}: Erasing the entire chip and writing two pages ...");
+    run_flash_erase(&mut session, &progress, true)?;
+    // TODO: The sector used here is not necessarily the sector the flash algorithm targets.
+    // Make this configurable.
+    let mut readback = vec![0; flash_properties.sectors[0].size as usize];
+    session.core(0)?.read_8(0x0, &mut readback)?;
+    assert!(
+        !readback.iter().any(|v| *v != erased_state),
+        "Not all sectors were erased"
+    );
+
+    let mut loader = session.target().flash_loader();
+    let data = (0..data_size)
+        .into_iter()
+        .map(|n| (n % 256) as u8)
+        .collect::<Vec<_>>();
+    loader.add_data(0x1, &data)?;
+    run_flash_download(&mut session, loader, &progress, true)?;
+    let mut readback = vec![0; data_size as usize];
+    session.core(0)?.read_8(0x1, &mut readback)?;
+    assert_eq!(readback, data);
+
+    println!("{test}: Erasing sectorwise and writing two pages double buffered ...");
+    run_flash_erase(&mut session, &progress, false)?;
+    // TODO: The sector used here is not necessarily the sector the flash algorithm targets.
+    // Make this configurable.
+    let mut readback = vec![0; flash_properties.sectors[0].size as usize];
+    session.core(0)?.read_8(0x0, &mut readback)?;
+    assert!(
+        !readback.iter().any(|v| *v != erased_state),
+        "Not all sectors were erased"
+    );
+
+    let mut loader = session.target().flash_loader();
+    let data = (0..data_size)
+        .into_iter()
+        .map(|n| (n % 256) as u8)
+        .collect::<Vec<_>>();
+    loader.add_data(0x1, &data)?;
+    run_flash_download(&mut session, loader, &progress, false)?;
+    let mut readback = vec![0; data_size as usize];
+    session.core(0)?.read_8(0x1, &mut readback)?;
+    assert_eq!(readback, data);
+
+    Ok(())
+}
+
+/// Performs the flash download with the given loader. Ensure that the loader has the data to load already stored.
+/// This function also manages the update and display of progress bars.
+pub fn run_flash_download(
+    session: &mut Session,
+    loader: FlashLoader,
+    progress: &FlashProgress,
+    disable_double_buffering: bool,
+) -> Result<()> {
+    let mut download_option = DownloadOptions::default();
+    download_option.keep_unwritten_bytes = false;
+    download_option.disable_double_buffering = disable_double_buffering;
+
+    download_option.progress = Some(progress);
+    download_option.skip_erase = true;
+
+    loader.commit(session, download_option)?;
+
+    Ok(())
+}
+
+/// Erases the flash.
```

Comment:
This is only true if `do_chip_erase` is true, otherwise it will just erase two sectors.